File: pipeline/optimizer.py
"""Reference survivor-path optimizer (Python).

The browser runs a JavaScript port of this; ``tests/test_parity.py`` checks
the two agree on the same data.

Objective for a path is the sum over weeks of
``log p  −  λ · log fs``
where ``p`` is the pick's win probability and ``fs`` is the expected fraction
of the field that survives the week *given that pick wins*
(``share_pick + Σ_{other} share·p``).  λ = 0 maximizes pure win-out
probability; λ = 1 maximizes the product of weekly pool "expected value"
(SurvivorGrid's EV), which rewards low-popularity picks.
"""
from __future__ import annotations

import math

import numpy as np
from scipy.optimize import linear_sum_assignment

BIG = 1e6


def field_survival(week_games: list[dict], pick: str) -> float:
    """Expected fraction of the field surviving this week if ``pick`` wins."""
    shares = {g["team"]: (g.get("pick") or 0.0) for g in week_games}
    total = sum(shares.values())
    if total <= 0:
        return 1.0
    shares = {t: s / total for t, s in shares.items()}
    fs = shares.get(pick, 0.0)
    for g in week_games:
        if g["team"] != pick:
            fs += shares[g["team"]] * g["p"]
    return max(fs, 1e-6)


def build_cost(games: list[dict], weeks: list[int], teams: list[str], lam: float) -> np.ndarray:
    by_week = {w: [g for g in games if g["week"] == w] for w in weeks}
    cost = np.full((len(weeks), len(teams)), BIG)
    for i, w in enumerate(weeks):
        for g in by_week[w]:
            if g["team"] in teams:
                j = teams.index(g["team"])
                p = min(max(g["p"], 1e-4), 1 - 1e-4)
                c = -math.log(p)
                if lam > 0:
                    c += lam * math.log(field_survival(by_week[w], g["team"]))
                cost[i, j] = c
    return cost


def best_path(games: list[dict], used: set[str], weeks: list[int], lam: float = 0.0,
              force: str | None = None) -> list[tuple[int, str]] | None:
    teams = sorted({g["team"] for g in games} - set(used))
    if len(weeks) > len(teams):
        return None
    cost = build_cost(games, weeks, teams, lam)
    if force is not None:
        if force not in teams or cost[0, teams.index(force)] >= BIG:
            return None
        j = teams.index(force)
        cost[0, :] = BIG
        cost[0, j] = -1e3   # cheap enough to always be chosen, finite so the solver stays stable
    rows, cols = linear_sum_assignment(cost)
    path = [(weeks[i], teams[j]) for i, j in zip(rows, cols)]
    if any(cost[i, j] >= BIG for i, j in zip(rows, cols)):
        return None
    return sorted(path)


def path_win_out(games: list[dict], path: list[tuple[int, str]]) -> float:
    idx = {(g["week"], g["team"]): g["p"] for g in games}
    return float(np.prod([idx[(w, t)] for w, t in path]))

File: docs/js/optimizer.js
/**
 * Survivor path optimizer. Port of pipeline/optimizer.py; keep the two in sync.
 *
 * Path objective: Σ_weeks [ log p − λ · log fs ], where fs is the expected
 * fraction of the field that survives the week given the pick wins.
 * λ = 0 → maximize win-out probability; λ = 1 → maximize weekly pool EV.
 */
import { assign, BIG } from './hungarian.js';

export function indexByWeek(games) {
  const by = new Map();
  for (const g of games) {
    if (g.result != null) continue;           // completed games are not pickable
    if (!by.has(g.week)) by.set(g.week, []);
    by.get(g.week).push(g);
  }
  return by;
}

export function fieldSurvival(weekGames, pick) {
  let total = 0;
  for (const g of weekGames) total += g.pick || 0;
  if (total <= 0) return 1;
  let fs = 0;
  for (const g of weekGames) {
    const share = (g.pick || 0) / total;
    fs += g.team === pick ? share : share * g.p;
  }
  return Math.max(fs, 1e-6);
}

function clampP(p) { return Math.min(Math.max(p, 1e-4), 1 - 1e-4); }

export function buildCost(byWeek, weeks, teams, lam) {
  const col = new Map(teams.map((t, j) => [t, j]));
  const cost = weeks.map(() => new Float64Array(teams.length).fill(BIG));
  weeks.forEach((w, i) => {
    const wg = byWeek.get(w) || [];
    for (const g of wg) {
      const j = col.get(g.team);
      if (j === undefined) continue;
      let c = -Math.log(clampP(g.p));
      if (lam > 0) c += lam * Math.log(fieldSurvival(wg, g.team));
      cost[i][j] = c;
    }
  });
  return cost;
}

/**
 * @returns {Array<{week:number, team:string}>|null}
 */
export function bestPath(byWeek, allTeams, used, weeks, lam = 0, force = null) {
  const usedSet = new Set(used);
  const teams = allTeams.filter((t) => !usedSet.has(t)).sort();
  if (weeks.length > teams.length) return null;
  const cost = buildCost(byWeek, weeks, teams, lam);
  if (force !== null) {
    const j = teams.indexOf(force);
    if (j < 0 || cost[0][j] >= BIG) return null;
    cost[0].fill(BIG);
    cost[0][j] = -1e3;
  }
  const a = assign(cost);
  const path = [];
  for (let i = 0; i < weeks.length; i++) {
    if (cost[i][a[i]] >= BIG) return null;
    path.push({ week: weeks[i], team: teams[a[i]] });
  }
  return path;
}

export function pathWinOut(byWeek, path) {
  let p = 1;
  for (const { week, team } of path) {
    const g = (byWeek.get(week) || []).find((x) => x.team === team);
    p *= g ? g.p : 0;
  }
  return p;
}
